In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('../datas/titanic/train.csv', sep = ',')

read_csv에서는 sep 인자 값이 있다. default는 , 이지만, \t 처럼 tab을 기준으로 적용시킬 수도 있다.

In [3]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
data.shape

(891, 12)

dataframe은 데이터뿐만 아니라 칼럼의 타입, null 데이터 개수, 데이터 분포도 등의 메타 데이터 등도 조회가 가능하다.

**info()와 describe()로 정보 조회가 가능하다. 먼저 info**


In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


890개의 row가 있다는 것과, 12개의 컬럼이 있다고 보여준다.

또한, 데이터 타입에 대해서도 알려주고 있다(int64 등)

그리고 중간에 891, 714 등의 숫자를 표시해주는 col이 있다. 전체 891 entries라고 나와 있는데 891 이라는 것은 데이터가 온전하다는 것이다.  
그러나 Age를 보면 714로 되어 있다. 이 말은 즉, 891 - 714 = 177 177개만큼 null 데이터가 있다는 것이다!

**describe는 칼럼별 숫자형 데이터 값의 n-percentile 분포도, 평균값, 최댓값, 최솟값을 나타낸다**

In [6]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

count : not null인 데이터 건수  
mean : 데이터의 평균값  
std : 표준편차  
min, max : 최소, 최댓값  
25%, 50%, 75% : 25 percentile 값, 50 percentile 값, 75 percentile 값  

survived를 보니 max가 1이고 min이 0이다. 이것을 보면 1과 0으로 이루어졌다고 추측할 수 있다.  
Pclass 경우도 mean이 2.3, max가 3, min이 1인 것을 보아하니 1~3 값 정도로 이루어졌다고 추측할 수 있다.

In [7]:
pclass = data['Pclass'].value_counts()
pclass

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [8]:
type(pclass)

pandas.core.series.Series

**series는 index와 단 하나의 칼럼으로 이루어진 데이터 셋트이다**

# numpy, 딕셔너리(dict), list과 dataframe 상호 변환

In [9]:
# 1차원 numpy -> dataframe
col = ['col1']
array = np.array([1,2,3])
d = pd.DataFrame(array, columns = col)
d

col1
0     1
1     2
2     3

In [10]:
# 2차원 numpy -> dataframe
col = ['col1', 'col2', 'col3']
array = np.array([[1,2,3],
                 [4,5,6]])

d = pd.DataFrame(array, columns = col)
d

col1  col2  col3
0     1     2     3
1     4     5     6

In [11]:
#딕셔너리 -> dataframe
d = {
    'col1' : [1,2],
    'col2' : [3,4],
    'col3' : [5,6]
}
d = pd.DataFrame(d)
d

col1  col2  col3
0     1     3     5
1     2     4     6

In [12]:
# dataframe -> list, dict

array = d.values
print(type(array), '\n',array)

<class 'numpy.ndarray'> 
 [[1 3 5]
 [2 4 6]]


In [13]:
array = d.values.tolist()
print(type(array),'\n', array)

<class 'list'> 
 [[1, 3, 5], [2, 4, 6]]


In [14]:
dic = d.to_dict('list')
dic

{'col1': [1, 2], 'col2': [3, 4], 'col3': [5, 6]}

In [15]:
data['Age_mul_10'] = data['Age'] * 10
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_mul_10  
0      0         A/5 21171   7.2500   NaN        S       220.0  
1      0          PC 17599  71.2833   C85        C       380.0  
2      0  STON/O2. 3101282   7.9250   NaN        S       260.0  
3      0            113803  53.1000  C123        S       350.0  
4      0            373450   8.0500   NaN        S       350.0

In [16]:
data['Age_mul_10'] = data['Age_mul_10'] + 100
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_mul_10  
0      0         A/5 21171   7.2500   NaN        S       320.0  
1      0          PC 17599  71.2833   C85        C       480.0  
2      0  STON/O2. 3101282   7.9250   NaN        S       360.0  
3      0            113803  53.1000  C123        S       450.0  
4      0            373450   8.0500   NaN        S       450.0

# dataframe 데이터 삭제

drop 메소드를 이용한다. 조금 혼동이 될 수 있기에 주의가 필요하다.

drop() 메소드는 인자값이 여러개가 있다. axis, inplace, labels가 대표적인 인자 값이다.

axis는 축이다. 0은 행 방향의 축, 1은 열 방향의 축이다.

만약, 컬럼을 삭제하고 싶으면 axis = 1을 하면 된다.데이터를 삭제하고 싶으면 axis = 0으로

1로 하면 labels에는 지정된 컬럼이 삭제가 되고 0으로 하면 index로 자동으로 인지한다.

inplace는 기존 dataframe에 적용을 시킬 것인지 여부를 따진다. 기본은 False이다.



In [17]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_mul_10  
0      0         A/5 21171   7.2500   NaN        S       320.0  
1      0          PC 17599  71.2833   C85        C       480.0  
2      0  STON/O2. 3101282   7.9250   NaN        S       360.0  
3      0            113803  53.1000  C123        S       450.0  
4      0            373450   8.0500   NaN        S       450.0

In [18]:
data['test'] = 0

In [19]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_mul_10  test  
0      0         A/5 21171   7.2500   NaN        S       320.0     0  
1      0          PC 17599  71.2833   C85        C       480.0     0  
2      0  STON/O2. 3101282   7.9250   NaN        S       360.0     0  
3      0            113803  53.1000  C123        S       450.0     0  
4      0            373450   8.0500   NaN        S       450.0     0

In [20]:
data = data.drop('Age_mul_10', axis = 1)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  test  
0      0         A/5 21171   7.2500   NaN        S     0  
1      0          PC 17599  71.2833   C85        C     0  
2      0  STON/O2. 3101282   7.9250   NaN        S     0  
3      0            113803  53.1000  C123        S     0  
4      0            373450   8.0500   NaN        S     0

In [21]:
test = data.drop('test', axis = 1, inplace = True)
print(test)

None


In [22]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [23]:
data['test1'] = 0
data['test2'] = 0
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  test1  test2  
0      0         A/5 21171   7.2500   NaN        S      0      0  
1      0          PC 17599  71.2833   C85        C      0      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0      0  
3      0            113803  53.1000  C123        S      0      0  
4      0            373450   8.0500   NaN        S      0      0

In [24]:
data.drop(['test1', 'test2'], axis = 1, inplace = True)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [25]:
data_copy = data.copy()
data_copy.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [26]:
data_copy.drop([1,2,3], axis = 0, inplace = True)
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
4            5         0       3        Allen, Mr. William Henry  male  35.0   
5            6         0       3                Moran, Mr. James  male   NaN   
6            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
7            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
4      0      0     373450   8.0500   NaN        S  
5      0      0     330877   8.4583   NaN        Q  
6      0      0      17463  51.8625   E46        S  
7      3      1     349909  21.0750   NaN        S

# index 객체

레코드를 유일하게 식별해주는 객체이다.

dataframe, series에서 inde 객체만 추출하려면 DataFrame.index 또는 Series.index를 하면 된다.

In [27]:
indexs = data.index
print(indexs)
print(type(indexs))
print(type(indexs.values))
print("리스트 출력 : \n ", indexs.values[:10])

RangeIndex(start=0, stop=891, step=1)
<class 'pandas.core.indexes.range.RangeIndex'>
<class 'numpy.ndarray'>
리스트 출력 : 
  [0 1 2 3 4 5 6 7 8 9]


**인덱스 값은 함부로 변경 불가**

In [28]:
# indexs[0] = 5

**Series 또한 index 객체가 존재하며 sum, min 등 연산 함수를 적용할 때 index는 제외시키고 진행한다**

In [29]:
data_fare = data['Fare']
print(type(data_fare))

<class 'pandas.core.series.Series'>


In [30]:
print(data_fare.index.values[:10])
print("max : ", data_fare.max())
print("sum : ", data_fare.sum())

[0 1 2 3 4 5 6 7 8 9]
max :  512.3292
sum :  28693.9493


**reset_index() 메서드를 수행하면 새롭게 인덱스를 숫자형으로 만들고 기존 인덱스를 컬름으로 새롭게 넣어준다**

여기에는 inplace가 마찬가지로 있는데 True는 기존 dataframe에 적용, False는 반환을 한다

In [31]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
4            5         0       3        Allen, Mr. William Henry  male  35.0   
5            6         0       3                Moran, Mr. James  male   NaN   
6            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
7            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
4      0      0     373450   8.0500   NaN        S  
5      0      0     330877   8.4583   NaN        Q  
6      0      0      17463  51.8625   E46        S  
7      3      1     349909  21.0750   NaN        S

In [32]:
data_copy2 = data_copy.reset_index(inplace = False)
data_copy2.head()

index  PassengerId  Survived  Pclass                            Name   Sex  \
0      0            1         0       3         Braund, Mr. Owen Harris  male   
1      4            5         0       3        Allen, Mr. William Henry  male   
2      5            6         0       3                Moran, Mr. James  male   
3      6            7         0       1         McCarthy, Mr. Timothy J  male   
4      7            8         0       3  Palsson, Master. Gosta Leonard  male   

    Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
0  22.0      1      0  A/5 21171   7.2500   NaN        S  
1  35.0      0      0     373450   8.0500   NaN        S  
2   NaN      0      0     330877   8.4583   NaN        Q  
3  54.0      0      0      17463  51.8625   E46        S  
4   2.0      3      1     349909  21.0750   NaN        S

이렇게 기존 인덱스가 연속된 숫자형이 아닐 때 사용하면 매우 유용하다!

근데 만약 저 컬럼을 추가시키지 않고 단순히 index만 초기화 하고 싶으면??

drop 인자를 이용한다.

In [33]:
data_copy2 = data_copy.reset_index(inplace=False, drop=True)
data_copy2.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
1            5         0       3        Allen, Mr. William Henry  male  35.0   
2            6         0       3                Moran, Mr. James  male   NaN   
3            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
4            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
1      0      0     373450   8.0500   NaN        S  
2      0      0     330877   8.4583   NaN        Q  
3      0      0      17463  51.8625   E46        S  
4      3      1     349909  21.0750   NaN        S

# pandas 데이터 셀렉션, 필터링



In [34]:
data['Pclass'].head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [35]:
data[['Pclass', 'Sex']].head()

Pclass     Sex
0       3    male
1       1  female
2       3  female
3       1  female
4       3    male

In [36]:
# data[0] 이렇게는 안된다.

In [37]:
# 이렇게는 가능하다
data[ 0:2 ]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

**불린 인덱스(boolean index)**

In [38]:
data[data['Pclass'] == 3].head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

**loc, iloc를 이용해 데이터를 선택한다**

loc : 명칭 기반 인덱싱  
iloc : 위치 기반 인덱싱

ix : 위치, 명칭을 혼합하여 사용가능하지만 사용을 추천하지 않는다. 삭제될 예정이라고 함

In [39]:
data.ix[0,'Pclass']

d:\anaconda3\envs\soojin\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


3

In [40]:
data.ix[0:2, ['Pclass', 'Sex']]

d:\anaconda3\envs\soojin\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.
d:\anaconda3\envs\soojin\lib\site-packages\pandas\core\indexing.py:822: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)


Pclass     Sex
0       3    male
1       1  female
2       3  female

In [41]:
# loc 이용

data.loc[0:5, ['Pclass', 'Sex']]

Pclass     Sex
0       3    male
1       1  female
2       3  female
3       1  female
4       3    male
5       3    male

In [42]:
#iloc 이용
data.iloc[0:5, [2, 5]]

Pclass   Age
0       3  22.0
1       1  38.0
2       3  26.0
3       1  35.0
4       3  35.0

**이 둘의 차이점은 바로 인덱스 범위**

loc는 명칭기반이라서 0:5의 뜻은 인덱스 0부터 인덱스 5까지의 값을 뜻한다. 그래서 0 ~ 5의 값이 출력이 되므로 6개가 출력  
iloc는 위치 기반이라서 0:5의 뜻은 5번 인덱스 전까지의 값을 뜻한다. 그래서 0 ~ 4의 값이 출력된다.  

**불린 인덱싱 (boolean indexing)**  

데이터 필터링 방식이다. 조건에 의해서 데이터를 가지고 올 수 있다.

In [43]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
4            5         0       3        Allen, Mr. William Henry  male  35.0   
5            6         0       3                Moran, Mr. James  male   NaN   
6            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
7            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
4      0      0     373450   8.0500   NaN        S  
5      0      0     330877   8.4583   NaN        Q  
6      0      0      17463  51.8625   E46        S  
7      3      1     349909  21.0750   NaN        S

In [44]:
data_age_60 = data_copy[data_copy['Age'] > 60]
data_age_60.head()

PassengerId  Survived  Pclass                            Name   Sex  \
33            34         0       2           Wheadon, Mr. Edward H  male   
54            55         0       1  Ostby, Mr. Engelhart Cornelius  male   
96            97         0       1       Goldschmidt, Mr. George B  male   
116          117         0       3            Connors, Mr. Patrick  male   
170          171         0       1       Van der hoef, Mr. Wyckoff  male   

      Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
33   66.0      0      0  C.A. 24579  10.5000   NaN        S  
54   65.0      0      1      113509  61.9792   B30        C  
96   71.0      0      0    PC 17754  34.6542    A5        C  
116  70.5      0      0      370369   7.7500   NaN        Q  
170  61.0      0      0      111240  33.5000   B19        S

In [45]:
data_age_60.shape

(22, 12)

특정 컬럼만 뽑아낼 수도 있다.

In [46]:
data_copy[data_copy['Age'] > 60][['Name', 'Age']].head()

Name   Age
33            Wheadon, Mr. Edward H  66.0
54   Ostby, Mr. Engelhart Cornelius  65.0
96        Goldschmidt, Mr. George B  71.0
116            Connors, Mr. Patrick  70.5
170       Van der hoef, Mr. Wyckoff  61.0

혹은 loc를 적용시켜서 뺄 수도 있다.

In [47]:
data_copy.loc[data_copy['Age'] > 60, ['Name', 'Age']].head()

Name   Age
33            Wheadon, Mr. Edward H  66.0
54   Ostby, Mr. Engelhart Cornelius  65.0
96        Goldschmidt, Mr. George B  71.0
116            Connors, Mr. Patrick  70.5
170       Van der hoef, Mr. Wyckoff  61.0

즉, 행에서 age > 60인 애들을 뽑아내고 그 중 컬럼을 name, age만 뽑아내는 것이다.

and 조건은 &  
or 조건은 |  
not 조건은 ~  
을 이용해서 복합적으로 뽑아낼 수 있다.

In [48]:
and_data = data_copy[ (data_copy['Age'] > 60) & (data_copy['Pclass'] == 1) ]
and_data.head()

PassengerId  Survived  Pclass                               Name     Sex  \
54            55         0       1     Ostby, Mr. Engelhart Cornelius    male   
96            97         0       1          Goldschmidt, Mr. George B    male   
170          171         0       1          Van der hoef, Mr. Wyckoff    male   
252          253         0       1          Stead, Mr. William Thomas    male   
275          276         1       1  Andrews, Miss. Kornelia Theodosia  female   

      Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
54   65.0      0      1    113509  61.9792   B30        C  
96   71.0      0      0  PC 17754  34.6542    A5        C  
170  61.0      0      0    111240  33.5000   B19        S  
252  62.0      0      0    113514  26.5500   C87        S  
275  63.0      1      0     13502  77.9583    D7        S

In [49]:
and_data.shape

(14, 12)

**loc도 적용 가능**

In [50]:
and_data = data_copy.loc[(data_copy['Age'] > 60 ) & (data_copy['Pclass'] == 1), :]
and_data.head()

PassengerId  Survived  Pclass                               Name     Sex  \
54            55         0       1     Ostby, Mr. Engelhart Cornelius    male   
96            97         0       1          Goldschmidt, Mr. George B    male   
170          171         0       1          Van der hoef, Mr. Wyckoff    male   
252          253         0       1          Stead, Mr. William Thomas    male   
275          276         1       1  Andrews, Miss. Kornelia Theodosia  female   

      Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
54   65.0      0      1    113509  61.9792   B30        C  
96   71.0      0      0  PC 17754  34.6542    A5        C  
170  61.0      0      0    111240  33.5000   B19        S  
252  62.0      0      0    113514  26.5500   C87        S  
275  63.0      1      0     13502  77.9583    D7        S

In [51]:
and_data.shape

(14, 12)

아래와 같이 활용도 가능

In [52]:
cond1 = data_copy['Age'] > 60
cond2 = data_copy['Pclass'] == 1
data_copy[cond1 & cond2].head()

PassengerId  Survived  Pclass                               Name     Sex  \
54            55         0       1     Ostby, Mr. Engelhart Cornelius    male   
96            97         0       1          Goldschmidt, Mr. George B    male   
170          171         0       1          Van der hoef, Mr. Wyckoff    male   
252          253         0       1          Stead, Mr. William Thomas    male   
275          276         1       1  Andrews, Miss. Kornelia Theodosia  female   

      Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
54   65.0      0      1    113509  61.9792   B30        C  
96   71.0      0      0  PC 17754  34.6542    A5        C  
170  61.0      0      0    111240  33.5000   B19        S  
252  62.0      0      0    113514  26.5500   C87        S  
275  63.0      1      0     13502  77.9583    D7        S

# Aggregation 함수, GroupBy 적용하기

**정렬하기**

dataframe과 series 정렬을 위해서는 sort_values() 메소드를 이용한다. RDBMS의 order by와 비슷하다.  
sort_values에는 inplace, by, ascending의 입력 파라미터 값이 있다.  
by : 특정 컬럼으로 정렬  
ascending : 기본은 True. False로 하면 내림차순 정렬  
inplace : True로 하면 원본 데이터에 적용  


In [53]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
4            5         0       3        Allen, Mr. William Henry  male  35.0   
5            6         0       3                Moran, Mr. James  male   NaN   
6            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
7            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
4      0      0     373450   8.0500   NaN        S  
5      0      0     330877   8.4583   NaN        Q  
6      0      0      17463  51.8625   E46        S  
7      3      1     349909  21.0750   NaN        S

In [54]:
data_copy.sort_values(by=['Age']).head()

PassengerId  Survived  Pclass                             Name     Sex  \
803          804         1       3  Thomas, Master. Assad Alexander    male   
755          756         1       2        Hamalainen, Master. Viljo    male   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
644          645         1       3           Baclini, Miss. Eugenie  female   
831          832         1       2  Richards, Master. George Sibley    male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
803  0.42      0      1    2625   8.5167   NaN        C  
755  0.67      1      1  250649  14.5000   NaN        S  
469  0.75      2      1    2666  19.2583   NaN        C  
644  0.75      2      1    2666  19.2583   NaN        C  
831  0.83      1      1   29106  18.7500   NaN        S

In [55]:
data_copy.sort_values(by=['Age'],ascending = False).head()

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
851          852         0       3                   Svensson, Mr. Johan   
96            97         0       1             Goldschmidt, Mr. George B   
493          494         0       1               Artagaveytia, Mr. Ramon   
116          117         0       3                  Connors, Mr. Patrick   

      Sex   Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
630  male  80.0      0      0     27042  30.0000   A23        S  
851  male  74.0      0      0    347060   7.7750   NaN        S  
96   male  71.0      0      0  PC 17754  34.6542    A5        C  
493  male  71.0      0      0  PC 17609  49.5042   NaN        C  
116  male  70.5      0      0    370369   7.7500   NaN        Q

In [56]:
data_copy.sort_values(by=['Age', 'SibSp']).head()

PassengerId  Survived  Pclass                             Name     Sex  \
803          804         1       3  Thomas, Master. Assad Alexander    male   
755          756         1       2        Hamalainen, Master. Viljo    male   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
644          645         1       3           Baclini, Miss. Eugenie  female   
78            79         1       2    Caldwell, Master. Alden Gates    male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
803  0.42      0      1    2625   8.5167   NaN        C  
755  0.67      1      1  250649  14.5000   NaN        S  
469  0.75      2      1    2666  19.2583   NaN        C  
644  0.75      2      1    2666  19.2583   NaN        C  
78   0.83      0      2  248738  29.0000   NaN        S

**Aggregation 함수 적용**

min(), max(), sum(), count()와 같은 aggregation 함수 적용은 RDBMS SQL의 aggregation 함수 적용과 유사하다.  
하지만 dataframe의 경우 바로 aggregation을 호출할 경우 모든 칼럼에 agg를 적용한다는 특징이 있다.  
아래와 같이.

In [57]:
data_copy.count()

PassengerId    888
Survived       888
Pclass         888
Name           888
Sex            888
Age            711
SibSp          888
Parch          888
Ticket         888
Fare           888
Cabin          202
Embarked       886
dtype: int64

특정 컬럼에 적용하려면 아래와 같이 진행한다.

In [58]:
data_copy[['Age', 'SibSp']].mean()

Age      29.685190
SibSp     0.522523
dtype: float64

**groupby() 적용**

dataframe의 groupby는 RDBMS SQL의 groupby와 유사하면서도 다르다.  
groupby 사용 시 입력 파라미터로 by를 기준으로 그룹을 묶어준다.

In [59]:
data_copy_group = data_copy.groupby(by='Pclass')
data_copy_group

이렇게 groupby를 하면 DataFrameGroupBy라는 객체가 반환되어서 생성된다.  
SQL과 다르게 groupby에 적용된 객체에 aggregation을 적용하면 모든 컬럼에 적용한다.

In [60]:
data_copy_group.count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               214       214   214  214  184    214    214     214   214   
2               184       184   184  184  173    184    184     184   184   
3               490       490   490  490  354    490    490     490   490   

        Cabin  Embarked  
Pclass                   
1         174       212  
2          16       184  
3          12       490

만약, 이것을 SQL로 했다면 SELECT COUNT(PassengerId), COUNT(Survived).. FROM data_copy GROUP BY Pclass 등으로 해야 했을 것이다.  

특정 컬럼에 agg를 적용하고 싶으면 특정 컬럼만 빼서 하면 된다.

In [61]:
data_copy_group['Fare'].max()

Pclass
1    512.3292
2     73.5000
3     69.5500
Name: Fare, dtype: float64

In [62]:
data_copy_group[['Fare', 'Age']].count()

Fare  Age
Pclass           
1        214  184
2        184  173
3        490  354

그리고 SQL에서 SELECT MAX(Age), MIN(Age) FROM data_copy GROUP BY Pclass와 같은 SQL을 pandas groupby에서는 아래와 같이 할 수 있다.

In [63]:
data_copy.groupby(by='Pclass')['Age'].agg([min, max])

min   max
Pclass            
1       0.92  80.0
2       0.67  70.0
3       0.42  74.0

그러면 여러개의 컬럼에 적용은 어떻게 시킬까?  
예를 들어서 SELECT MAX(Age), SUM(Fare) FROM data_copy GROUP BY Pclass와 같은 SQL을 만드려면?

In [64]:
agg_d = {
    'Age' : 'max',
    'Fare' : 'count'
}

data_copy.groupby(by='Pclass').agg(agg_d)

Age  Fare
Pclass            
1       80.0   214
2       70.0   184
3       74.0   490

# 결손데이터 처리하기


In [65]:
data_copy.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          686
Embarked         2
dtype: int64

In [66]:
data_copy['Cabin'] = data_copy['Cabin'].fillna('C123')
data_copy.head(10)

PassengerId  Survived  Pclass  \
0             1         0       3   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   

    Parch     Ticket     Fare Cabin Embarked  
0       0  A/5 21171   7.2500  C123        S  
4       0     373450   8.0500  C123        S  
5       0     330877   8.4583  C123        Q  
6       0      17463  51.8625   E46        S  
7       1     349909  21.0750  C123        S  
8       2     347742  11.1333  C123        S  
9       0     237736  30.0708  C123        C  
10      1    PP 9549  16.7000    G6        S  
11      0     113783  26.5500  C103        S  
12      0  A/5. 2151   8.0500  C123        S

**fillna의 inplace도 True하면 원본 데이터에 바로 적용**


In [67]:
data_copy['Embarked'].fillna('S', inplace = True)
data_copy['Age'].fillna(data_copy['Age'].mean(), inplace=True)
data_copy.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

** lambda 사용 **

lamba x : x ** 2

x : 입력 인자  
x ** 2 : 계산 및 반환값

In [68]:
test = [1,2,3]
test = map(lambda x : x**2, test)
list(test)

[1, 4, 9]

In [69]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex  \
0            1         0       3         Braund, Mr. Owen Harris  male   
4            5         0       3        Allen, Mr. William Henry  male   
5            6         0       3                Moran, Mr. James  male   
6            7         0       1         McCarthy, Mr. Timothy J  male   
7            8         0       3  Palsson, Master. Gosta Leonard  male   

        Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
0  22.00000      1      0  A/5 21171   7.2500  C123        S  
4  35.00000      0      0     373450   8.0500  C123        S  
5  29.68519      0      0     330877   8.4583  C123        Q  
6  54.00000      0      0      17463  51.8625   E46        S  
7   2.00000      3      1     349909  21.0750  C123        S

In [70]:
data_copy['Name_len'] = data_copy['Name'].apply(lambda x : len(x))

In [71]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex  \
0            1         0       3         Braund, Mr. Owen Harris  male   
4            5         0       3        Allen, Mr. William Henry  male   
5            6         0       3                Moran, Mr. James  male   
6            7         0       1         McCarthy, Mr. Timothy J  male   
7            8         0       3  Palsson, Master. Gosta Leonard  male   

        Age  SibSp  Parch     Ticket     Fare Cabin Embarked  Name_len  
0  22.00000      1      0  A/5 21171   7.2500  C123        S        23  
4  35.00000      0      0     373450   8.0500  C123        S        24  
5  29.68519      0      0     330877   8.4583  C123        Q        16  
6  54.00000      0      0      17463  51.8625   E46        S        23  
7   2.00000      3      1     349909  21.0750  C123        S        30

if ~ else 구문 사용하기

In [72]:
data_copy['Age_c1'] = data_copy['Age'].apply(lambda x : 'child' if x <= 20 else 'adult')
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex  \
0            1         0       3         Braund, Mr. Owen Harris  male   
4            5         0       3        Allen, Mr. William Henry  male   
5            6         0       3                Moran, Mr. James  male   
6            7         0       1         McCarthy, Mr. Timothy J  male   
7            8         0       3  Palsson, Master. Gosta Leonard  male   

        Age  SibSp  Parch     Ticket     Fare Cabin Embarked  Name_len Age_c1  
0  22.00000      1      0  A/5 21171   7.2500  C123        S        23  adult  
4  35.00000      0      0     373450   8.0500  C123        S        24  adult  
5  29.68519      0      0     330877   8.4583  C123        Q        16  adult  
6  54.00000      0      0      17463  51.8625   E46        S        23  adult  
7   2.00000      3      1     349909  21.0750  C123        S        30  child

In [73]:
data_copy['Age_c2'] = data_copy['Age'].apply(lambda x : 'child' if x <= 17 else ('adult' if x <= 40 else 'older'))

In [74]:
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex  \
0            1         0       3         Braund, Mr. Owen Harris  male   
4            5         0       3        Allen, Mr. William Henry  male   
5            6         0       3                Moran, Mr. James  male   
6            7         0       1         McCarthy, Mr. Timothy J  male   
7            8         0       3  Palsson, Master. Gosta Leonard  male   

        Age  SibSp  Parch     Ticket     Fare Cabin Embarked  Name_len Age_c1  \
0  22.00000      1      0  A/5 21171   7.2500  C123        S        23  adult   
4  35.00000      0      0     373450   8.0500  C123        S        24  adult   
5  29.68519      0      0     330877   8.4583  C123        Q        16  adult   
6  54.00000      0      0      17463  51.8625   E46        S        23  adult   
7   2.00000      3      1     349909  21.0750  C123        S        30  child   

  Age_c2  
0  adult  
4  adult  
5  adult  
6  older  
7  child

In [75]:
data_copy['Age_c2'].value_counts()

adult    625
older    150
child    113
Name: Age_c2, dtype: int64

**별도 함수 적용**

In [76]:
def get_category(age):
    c = ''
    if age <= 5 : c = 'baby'
    elif age <= 12 : c = 'child'
    elif age <= 18 : c = 'teenager'
    elif age <= 25 : c = 'student'
    elif age <= 35 : c = 'young adult'
    elif age <= 60 : c = 'adult'
    else : c = 'elderly'
    return c

In [77]:
data_copy['Age_c3'] = data_copy['Age'].apply(lambda x : get_category(x))
data_copy.head()

PassengerId  Survived  Pclass                            Name   Sex  \
0            1         0       3         Braund, Mr. Owen Harris  male   
4            5         0       3        Allen, Mr. William Henry  male   
5            6         0       3                Moran, Mr. James  male   
6            7         0       1         McCarthy, Mr. Timothy J  male   
7            8         0       3  Palsson, Master. Gosta Leonard  male   

        Age  SibSp  Parch     Ticket     Fare Cabin Embarked  Name_len Age_c1  \
0  22.00000      1      0  A/5 21171   7.2500  C123        S        23  adult   
4  35.00000      0      0     373450   8.0500  C123        S        24  adult   
5  29.68519      0      0     330877   8.4583  C123        Q        16  adult   
6  54.00000      0      0      17463  51.8625   E46        S        23  adult   
7   2.00000      3      1     349909  21.0750  C123        S        30  child   

  Age_c2       Age_c3  
0  adult      student  
4  adult  young adult  
5  adult  young adult  
6  older        adult  
7  child         baby